### PR TITLE
Make optimization configurable

### PR DESCRIPTION
**Pull request by odespard**
Add a new ``optimization`` field in the config which can be used to specify the update rule parameters and the order of optimization.

### CODE REVIEW

File: alphadia/constants/default.yaml [hunk 1]
```diff
@@ -189,8 +189,23 @@ search_output:
   # can be either "parquet" or "tsv"
   file_format: "tsv"
 
+optimization:
+  order_of_optimization: null
```
Comment (mschwoer) [inline alphadia/constants/default.yaml:193]: please comment heavily here on what these parameters mean.

for `order_of_optimization`, state what is expected (list? integers? names?) and give an example (or set the default here)

File: alphadia/constants/default.yaml [hunk 2]
```diff
@@ -189,8 +189,23 @@ search_output:
...
+  ms2_error:
+      update_interval: 0.95
```
Comment (mschwoer) [inline alphadia/constants/default.yaml:195]: sure these are all correct? would have expected 0.99 and 1.1 here for some ..

File: alphadia/workflow/optimization.py [hunk 3]
```diff
@@ -508,9 +523,9 @@ def _propose_new_parameter(self, df: pd.DataFrame):
 
         """
 
-        return 1.1 * self.workflow.calibration_manager.get_estimator(
+        return self.update_factor * self.workflow.calibration_manager.get_estimator(
```
Comment (mschwoer) [inline alphadia/workflow/optimization.py:526]: now that we parametrized it, coming back to an earlier suggestion: could we have only one implementation of this?
(this would be a dedicated PR)

File: alphadia/workflow/peptidecentric.py [hunk 4]
```diff
@@ -323,32 +323,51 @@ def get_ordered_optimizers(self):
...
-        optimizers = [
-            ms2_optimizer,
-            rt_optimizer,
-            ms1_optimizer,
-            mobility_optimizer,
-        ]
-        targeted_optimizers = [
-            [
-                optimizer
+        if self.config["optimization"]["order_of_optimization"] is None:
```
Comment (mschwoer) [inline alphadia/workflow/peptidecentric.py:326]: the `if` branch could be removed if the `default.yaml` provided a default instead of `None`

**Review by mschwoer — COMMENTED** [review]
some suggestions for simplifications

File: alphadia/constants/default.yaml (hunk 2 above)
Comment (odespard) [reply alphadia/constants/default.yaml:195]: It depends on whether we want to allow distinct update rules for targeted and automatic calibration. I decided not to (since any given run will only use one of these) and, since the emphasis so far has been on preventing any performance degradation for the targeted case, I thought it best to use the default values for this case in the YAML.

File: alphadia/workflow/peptidecentric.py (hunk 4 above)
Comment (odespard) [reply alphadia/workflow/peptidecentric.py:326]: It couldn't be removed while maintaining precisely the same functionality, which is to ensure that targeted optimizers always come before automatic optimizers. Whether this functionality is worth maintaining is a separate question, but I thought it best to leave it in place for the time being as it seems a reasonable default.

File: alphadia/workflow/optimization.py (hunk 3 above)
Comment (odespard) [reply alphadia/workflow/optimization.py:526]: We could, and I suppose we could always overwrite the default method in a child class if needed. I'll make do this soon.

File: alphadia/constants/default.yaml (hunk 1 above)
Comment (odespard) [reply alphadia/constants/default.yaml:193]: I'll add comments. I am inclined not to set a default for the reasons set out in https://github.com/MannLabs/alphadia/pull/318#discussion_r1726472814.

Comment (GeorgWa) [reply alphadia/constants/default.yaml:197]: Please let me know if I have some missunderstanding here.

Should the optimiztion order be a debug/experimental config or something the user really wants to change?

If it's encouraged to be configured by users, the order should be independent of the question of target/automatic. for example `order_of_optimization: [('rt','mobility')('ms2')('ms1)]` 

If this is something experimental for us so we can test different combinations, let's mark it like this.

File: alphadia/constants/default.yaml (hunk 2 above)
Comment (GeorgWa) [reply alphadia/constants/default.yaml:203]: I assume this is experimental until we find something which always work?

I agree with Magnus, let's use different parameters, especially as they wont have the same metric update rules in the future. If we dont want to expose this in the gui, make it accessible, we don't ned to worry about having to many parameters.

File: alphadia/constants/default.yaml (hunk 1 above)
Comment (odespard) [reply alphadia/constants/default.yaml:197]: My view is that it's something experimental. How should I mark it as such?

Comment (mschwoer) [reply alphadia/constants/default.yaml:232]: just add a comment that this is only for expert users?

File: alphadia/workflow/peptidecentric.py (hunk 4 above)
Comment (mschwoer) [reply alphadia/workflow/peptidecentric.py:326]: I see .. should the "targeted optimizers always come before automatic optimizers." be added as a comment to the yaml?

**Review by mschwoer — APPROVED** [review]
LGTM

File: alphadia/workflow/peptidecentric.py (hunk 4 above)
Comment (odespard) [reply alphadia/workflow/peptidecentric.py:326]: OK.
